### PR TITLE
23350 - notice of withdrawal outputs - focus on NoW for existing businesses

### PR DESCRIPTION
**Pull request by eason-pan-bc**
*Issue #:* /bcgov/entity#23350

*Description of changes:*
- Added template for NoW
- Updated business details for NoW
- Updated receipt effective_date logic for NoW

**NoW Outputs Design**
https://www.figma.com/design/CLqwqGQYARaSrnzLwh6w8Q/Notice-of-Withdrawal?node-id=883-69288&p=f&t=U1sJ3PuPG1d604JJ-0

**Note** 
Since legal-api is not yet supporting filing NoW for temporary businesses, the verification done in this PR is for NoW filings for existing businesses. Temporary business part will be covered in https://github.com/bcgov/entity/issues/24988

#### Verification - Output Examples for Existing Businesses
- Receipt for NoW (No Fee): 
[NoW receipt.pdf](https://github.com/user-attachments/files/18234044/NoW.receipt.pdf)
- Receipt for NoW (using Routing Slip Number)
[NoW receipt - paid.pdf](https://github.com/user-attachments/files/18240964/NoW.receipt.-.paid.pdf)
- Receipt for businesses have a tax id (No Fee):
[NoW receipt with tax id.pdf](https://github.com/user-attachments/files/18234047/NoW.receipt.with.tax.id.pdf)
- NoW Output for Existing businesses:
[output for NoW - existing business.pdf](https://github.com/user-attachments/files/18234048/output.for.NoW.-.existing.business.pdf)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

**Comment by ArwenQin** [comment]
For the receipt, do we need to update the template to show the payment method, as in the design?
image

**Comment by eason-pan-bc** [comment]
> For the receipt, do we need to update the template to show the payment method, as in the design? image

it will show if there was a payment. I used the staff no payment option. That's why it's not showing in the sample.

**Comment by severinbeauvais** [comment]
Do you have a sample receipt, and application, for a withdrawn IA?

**Comment by eason-pan-bc** [comment]
> Do you have a sample receipt, and application, for a withdrawn IA?

Not at this moment, as the legal-api is not yet supporting filing for temporary businesses. And as mentioned above, this part will be covered in https://github.com/bcgov/entity/issues/24988

File: legal-api/report-templates/template-parts/common/businessDetails.html [hunk 1]
```diff
@@ -328,6 +328,28 @@
                {% endif %}
                <div class="pt-2">{{report_date_time}}</div>
             </td>
+         {% elif header.name == 'noticeOfWithdrawal' %}
+            <td class="col-33">
+               {% if not business or business.identifier.startswith('T') %}
+                  <div class="bold">Filed Date and Time:</div>
+               {% else %}
+                  <div class="bold">Incorporation Number:</div>
```
Comment (severinbeauvais) [inline legal-api/report-templates/template-parts/common/businessDetails.html:336]: You're not going to have an incorporation number (see also line 347) if the business identified starts with a T. Did Jacqueline want the T number displayed anyway?

See pink NOTES box:
image

**Review by severinbeauvais — APPROVED** [review]
LGTM.

~Any unit tests for this?~ _now added, thanks!_

File: legal-api/report-templates/template-parts/common/businessDetails.html (hunk 1 above)
Comment (eason-pan-bc) [reply legal-api/report-templates/template-parts/common/businessDetails.html:336]: The 'T' number won't be rendered, only if it's a business and identifier not starts with 'T' will render the incorporation number section. As you can see the logic in line 333 and 334.

**Comment by eason-pan-bc** [comment]
> LGTM.
> 
> Any unit tests for this?

Thank you for taking a look of this. I checked some previous PRs for updating or adding outputs, and I didn't see any new unit tests were added. So, I assume I don't need to add any?

File: legal-api/report-templates/template-parts/common/businessDetails.html (hunk 1 above)
Comment (eason-pan-bc) [reply legal-api/report-templates/template-parts/common/businessDetails.html:336]: Will double-check and verify this when doing the other ticket when supporting NoW for temporary businesses.

File: legal-api/src/legal_api/reports/report.py [hunk 2]
```diff
@@ -748,6 +751,18 @@ def _format_amalgamation_data(self, filing):
     def _format_certificate_of_amalgamation_data(self, filing):
         self._set_amalgamating_businesses(filing)
 
+    def _format_notice_of_withdrawal_data(self, filing):
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/reports/report.py:754]: This may not be enough for T business. Did you test this with T business?

Comment (eason-pan-bc) [reply legal-api/src/legal_api/reports/report.py:754]: Not at this moment, as the legal-api is not yet supporting filing NoW for temporary businesses. And as mentioned above, the T business part will be covered in https://github.com/bcgov/entity/issues/24988

File: legal-api/report-templates/template-parts/common/businessDetails.html (hunk 1 above)
Comment (severinbeauvais) [reply legal-api/report-templates/template-parts/common/businessDetails.html:336]: Oh, I saw the "not" but missed the "or". I think that logic is OK.

So, for a T business, the display is this?
- Filed Date and Time:
- Recognition Date and Time: (ie, effective) **maybe this should be a different label?**
- Retrieved Date and Time: (ie, report)

If OK then you can resolve this comment.

**Comment by severinbeauvais** [comment]
> > LGTM.
> > Any unit tests for this?
> 
> Thank you for taking a look of this. I checked some previous PRs for updating or adding outputs, and I didn't see any new unit tests were added. So, I assume I don't need to add any? And, this PR doesn't really have much back-end logic changes, it's mainly update / create report templates.

I might be wrong, but I thought there were unit tests to verify the data passed to the template...?

File: legal-api/report-templates/template-parts/common/businessDetails.html (hunk 1 above)
Comment (eason-pan-bc) [reply legal-api/report-templates/template-parts/common/businessDetails.html:336]: Yes, these 3 items for T business, follows the design. and will be verified later. 
image

and I've left a question to Jacqueline in the ticket for the expected value for the 'Recognition Date and Time', we will know when she's back. And I will update the details in the other tickets when it's clarified. 😃

**Comment by eason-pan-bc** [comment]
> > > LGTM.
> > > Any unit tests for this?
> > 
> > 
> > Thank you for taking a look of this. I checked some previous PRs for updating or adding outputs, and I didn't see any new unit tests were added. So, I assume I don't need to add any? And, this PR doesn't really have much back-end logic changes, it's mainly update / create report templates.
> 
> I might be wrong, but I thought there were unit tests to verify the data passed to the template...?

I think it's a good point, although I didn't see much, but let me add some unit tests for it.

**Review by ArwenQin — APPROVED** [review]
The templates look good

**Review by vysakh-menon-aot — APPROVED** [review]

File: legal-api/src/legal_api/reports/report.py [hunk 3]
```diff
@@ -748,6 +751,18 @@ def _format_amalgamation_data(self, filing):
...
+        withdrawn_filing_id = filing['noticeOfWithdrawal']['filingId']
+        withdrawn_filing = Filing.find_by_id(withdrawn_filing_id)
+        formatted_withdrawn_filing_type = self._format_filing_type(withdrawn_filing.filing_type)
+        filing['withdrawnFilingType'] = formatted_withdrawn_filing_type
+        withdrawn_filing_date = LegislationDatetime.as_legislation_timezone(withdrawn_filing.effective_date)
+        filing['withdrawnFilingEffectiveDate'] = LegislationDatetime.format_as_report_string(withdrawn_filing_date)
+
+    def _format_filing_type(self, filing_type_str: str):
+        words = ''.join(' ' + c if c.isupper() else c for c in filing_type_str).split()
+        return ' '.join(word.capitalize() if word != 'agm' else word.upper() for word in words)
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/reports/report.py:764]: This func is not required. Use `FilingMeta.get_display_name` instead https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/core/meta/filing.py#L783

Comment (eason-pan-bc) [reply legal-api/src/legal_api/reports/report.py]: Updated. Thanks for pointing it out.

**Pull request merged**